File: python/basics/hash_table/128_longest_consecutive_sequence.py
#Approach I: n log N => sort and then find the consecutive sequence
#check for the element which starts the sequence ... 
#O(n)
#space O(n)
def longest_consecutive_sequence(input:list[int]):
    hash_set = set(input)
    n = len(input)
    max_subseq_len=0
    for i in range(n):
        num=input[i-1]
        if num-1 not in hash_set:   #this is where the sequence starts... e.g 4,3,2... we start from 2. and go upwards
            subseq_len=1
            while num+1 in hash_set:
                subseq_len += 1
                num=num+1
            max_subseq_len = max(max_subseq_len, subseq_len)  

    return max_subseq_len




input = [100, 4, 200, 1,3,2]
result = longest_consecutive_sequence(input)
print("result=" + str(result))

input = [0,3,7,2,5,8,4,6,0,1]
result = longest_consecutive_sequence(input)
print("result=" + str(result))

File: problem_solving/hash_table/cpp/128_longest_consecutive_sequence.cpp
/*
#Approach I: n log N => sort and then find the consecutive sequence

#Approach II: insert the list into has set
#if the N has N-1 has set, then ignore.
    #if it does not, then keep looking for the numbers above it.
#check for the element which starts the sequence ... 
#O(n)
#space O(n)
#interesting
*/

#include<iostream>
#include<tuple>
#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"
#include<unordered_map>
using namespace std;


int longest_consecutive_sequence(const std::vector<int>& nums)  {
    std::unordered_set<int> sequence;
    for(int num : nums) {
        sequence.insert(num);
    }
    int max_len = 0;
    for(int num : sequence) {
        if (sequence.find(num-1) == sequence.end()) {
            int len = 1;
            while( sequence.find(num+1) != sequence.end() ) {
                len += 1;  
                num=num+1;  
            }
            max_len = std::max(max_len, len);
        }
    }
    return max_len;
}


int main() {
    std::vector<int> input;
    input = {100, 4, 200, 1,3,2};
    auto result = longest_consecutive_sequence(input);
    std::cout << "result=" << result << std::endl;
    //4

    input = {0,3,7,2,5,8,4,6,0,1};
    result = longest_consecutive_sequence(input);
    std::cout << "result=" << result << std::endl;
    //9

    input = {1,0,1,2};
    //Output: 3
    result = longest_consecutive_sequence(input);
    std::cout << "result=" << result << std::endl;
    //3
}


